compose/foundation/foundation/src/desktopMain/kotlin/androidx/compose/foundation/internal/ClipboardUtils.desktop.kt

```
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation.internal
+
+import androidx.compose.ui.platform.ClipEntry
+import androidx.compose.ui.text.AnnotatedString
+import java.awt.datatransfer.DataFlavor
+import java.awt.datatransfer.StringSelection
+import java.awt.datatransfer.UnsupportedFlavorException
+import java.io.IOException
+
+internal actual suspend fun ClipEntry.readText(): String? {
+    return try {
+        transferable.getTransferData(DataFlavor.stringFlavor) as String?
```

Maybe `as? String` instead? Shouldn't make a real difference because `DataFlavor.stringFlavor` should only return `String`s, but it looks a bit better. Also maybe check `isDataFlavorSupported` before calling `getTransferData` to avoid throwing an exception in a non-exceptional case.

But in general, we should document the behavior of the method in the `expect`. Was there a discussion about it? I mean whether it should throw an exception or return `null` when the exceptions below are thrown?

---

>But in general, we should document the behavior of the method in the expect. Was there a discussion about it? I mean whether it should throw an exception or return null when the exceptions below are thrown?

There were no such discussion.
The internal usages of this method do not expect any exceptions. 
I added the Kdoc for an expect. Also a ticket to upstream the doc: https://youtrack.jetbrains.com/issue/CMP-7544/Upstream-ClipboardUtils.kt-kdoc

---

Ah, right, this is internal only. Maybe it's worthwhile to catch all exceptions then, rather than the individual ones, with `runCatching`? Not 100% sure; sometimes it's better to crash so that a bug is caught and fixed faster.

---

I left only one exception which is expected - `IOException`. 
We can assume that other exceptions would mean there is a bug.